criu/files-reg.c

```diff
@@ -1384,30 +1384,42 @@ static int get_build_id_32(Elf32_Ehdr *file_header, unsigned char **build_id,
 	 * If the file has a build-id, it will be in the PT_NOTE program header
 	 * entry AKA the note sections.
 	 */
-	while (num_iterations-- && program_header <= program_header_end &&
-			program_header->p_type != PT_NOTE)
-		program_header++;
-
-	if (!num_iterations || program_header >= program_header_end) {
-		pr_warn("Couldn't find the note program header for file with fd %d\n", fd);
-		return -1;
-	}
+	while (num_iterations-- && program_header <= program_header_end) {
+		if (program_header->p_type != PT_NOTE) {
+			program_header++;
+			continue;
+		}
 
-	note_header = (Elf32_Nhdr *) (program_header->p_offset + (char *) file_header);
-	if (note_header <= (Elf32_Nhdr *) file_header)
-		return -1;
+		note_header = (Elf32_Nhdr *) (program_header->p_offset + (char *) file_header);
+		if (note_header <= (Elf32_Nhdr *) file_header) {
+			program_header++;
+			note_header = NULL;
+			continue;
+		}
 
-	note_header_end = (Elf32_Nhdr *) min_t(char*,
+		note_header_end = (Elf32_Nhdr *) min_t(char*,
 						(char *) note_header + program_header->p_filesz,
 						(char *) (file_header_end - sizeof(Elf32_Nhdr)));
 
-	/* The note type for the build-id is NT_GNU_BUILD_ID. */
-	while (note_header <= note_header_end && note_header->n_type != NT_GNU_BUILD_ID)
-		note_header = (Elf32_Nhdr *) ((char *) note_header + sizeof(Elf32_Nhdr) +
-						ALIGN(note_header->n_namesz, 4) +
-						ALIGN(note_header->n_descsz, 4));
+		/* The note type for the build-id is NT_GNU_BUILD_ID. */
+		while (note_header <= note_header_end && note_header->n_type != NT_GNU_BUILD_ID)
+			note_header = (Elf32_Nhdr *) ((char *) note_header + sizeof(Elf32_Nhdr) +
+							ALIGN(note_header->n_namesz, 4) +
+							ALIGN(note_header->n_descsz, 4));
+		
+		if (note_header >= note_header_end) {
+			program_header++;
+			note_header = NULL;
+			continue;
+		}
+	}
```

Comment:
you will go to the next loop iteration with the same  program_header. I think you need yo add break right before this line.

Reply:
I've added the break statement.